### ISSUE TITLE
remove `bidirectional` option from lm

### ISSUE THREAD

**Issue by flinder** [SMAPPNYU/stance_classification_experiment#20]
@yinleon could you take a look at this? 

You can reproduce the error by running `/ulmfit/example_notebooks/EXAMPLE_training_lm.ipynb` (make sure bidirectional is set to True).

Here's the traceback:
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-12-281bde31ef9f> in <module>
     27                         lm, loss, optimizer, scheduler, device,
     28                         train_dl, valid_dl,
---> 29                         best_model_path=model_file_lm)

~/smapp_stance_sentiment/ulmfit/training_utils.py in training_loop(batch_size, num_epochs, display_freq, model, criterion, optim, scheduler, device, training_set, validation_set, best_model_path, history)
    160                                             criterion=criterion, optim=optim,
    161                                             scheduler=scheduler, batch_size=batch_size,
--> 162                                             device=device)
    163             history.update_loss(loss)
    164 

~/smapp_stance_sentiment/ulmfit/training_utils.py in training_epoch(*args, **kwargs)
    129 def training_epoch(*args, **kwargs):
    130     '''Training Epoch'''
--> 131     return run_epoch(train=True, *args, **kwargs)
    132 
    133 

~/smapp_stance_sentiment/ulmfit/training_utils.py in run_epoch(model, dataset, criterion, optim, scheduler, batch_size, device, train, shuffle)
    107         # get a predition
    108         model.init_hidden()
--> 109         y_ = model(X)
    110 
    111         # calculate loss and accuracy

~/anaconda3/envs/ulmfit/lib/python3.6/site-packages/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    487             result = self._slow_forward(*input, **kwargs)
    488         else:
--> 489             result = self.forward(*input, **kwargs)
    490         for hook in self._forward_hooks.values():
    491             hook_result = hook(self, input, result)

~/smapp_stance_sentiment/ulmfit/model.py in forward(self, x, return_hidden, intermediaries)
    240         for l, (rnn, drop) in enumerate(zip(self.rnns,
    241                                             self.hidden_dropout)):
--> 242             output, hidden = rnn(output, self.hidden[l])
    243             new_hidden.append(hidden)
    244             raw_outputs.append(output)

~/anaconda3/envs/ulmfit/lib/python3.6/site-packages/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    487             result = self._slow_forward(*input, **kwargs)
    488         else:
--> 489             result = self.forward(*input, **kwargs)
    490         for hook in self._forward_hooks.values():
    491             hook_result = hook(self, input, result)

~/anaconda3/envs/ulmfit/lib/python3.6/site-packages/torch/nn/modules/rnn.py in forward(self, input, hx)
    173                 hx = (hx, hx)
    174 
--> 175         self.check_forward_args(input, hx, batch_sizes)
    176         _impl = _rnn_impls[self.mode]
    177         if batch_sizes is None:

~/anaconda3/envs/ulmfit/lib/python3.6/site-packages/torch/nn/modules/rnn.py in check_forward_args(self, input, hidden, batch_sizes)
    133             raise RuntimeError(
    134                 'input.size(-1) must be equal to input_size. Expected {}, got {}'.format(
--> 135                     self.input_size, input.size(-1)))
    136 
    137         if is_input_packed:

RuntimeError: input.size(-1) must be equal to input_size. Expected 300, got 600
```

**Comment by yinleon** [comment]
@flinder are the actual metrics worse, or just the sentence generation?
One is a problem with the model and one is probably a problem with `sample()`.

**Comment by flinder** [comment]
@yinleon, still working on it, but here some interesting intermediate (confusing?) results (also @vishakhpk in case you're looking at the model):

I just fixed our accuracy measure (see #22,  0ee2bb104a761c8bd95033424bc4284a974a29b0). As a baseline: Fastai produces ~27% categorical validation accuracy (predicting the next word correctly by assigning the highest 'probability' word from the output) after the second epoch (val loss ~4.3). All results are on the 25K imdb `training.csv`. 

Our model with `bidirectional=False` gets ~9% validation accuracy after 2 epochs (went up to ~15% after a few more, so this differential might be just a training problem we need to verify that)

Our model with `bidirectional=True` gets ~72%(!!) accuracy after a few epochs. But check the sample output:
```
Epoch: 0002; Train-Loss: 2.3952; Val-Loss 1.6140; Train-accuracy 0.5907; Val-Accuracy 0.7223
Sample: this is good better this this this this this this this this this this this this this this this this this this this this this this this this this this this this this this this this this this this this this this this this this this this this this this this this this
```
Validation loss is crazy too.

I think there's something going on with the sample function because when I run the following:
```
from training_utils import categorical_accuracy
it = iter(valid_dl)
X, y = next(it)
# Lm is the bidirectional model trained for a few epochs
lm.eval()
y_ = lm(X.to(device))
y = y.contiguous().view(-1).to(device)
categorical_accuracy(y_, y)
# This gives 0.73
y_pred = torch.max(y_, dim=1)[1]
# Translate predicitons to words
[vectorizer.idx2word[x.item()] for x in y_pred]
```
I get output like this: 
```
'a sudden action movie that i have ever seen . it was the first extent that i sat after watching it happens on the best horror in my couple , <UNK> . a entertaining between a opening crime and a great running reminded played well by david goldblum .
```
Which looks really good. I double checked the sampling function but couldn't find anything. Do you huys have an idea what could explain this?

But I feel like something must be wrong that gives us that crazy accuracy. It's not overfitting though, because the training loss is much smaller.
